### PR TITLE
[Jem] Adding UX/UI Options to Lists and the Profile Page

### PR DESCRIPTION
**Pull request by jemeduria**
## Context
- Users can create/delete song lists and add/remove songs to those lists. 

## Changes
- Created more API routes regarding editing of lists and contents
- Updated modals to disappear after use and added modals for better UX/UI functionality of lists
- Ensured that any updates made reflect on pages without refreshing.
- Added messages for missing or empty content

## Testing For UX/UI Updates
1. Create a new account.
2. Ensure that the no reviews, no albums, and no lists messages appear.
3. Navigate to the different sub-pages within a profile page (reviews, lists, albums). Ensure that the no reviews, no albums, and no lists messages appear.

## Testing For Lists
1. Create a new account.
5. Throughout all steps below: Check MongoDB to confirm that lists are being updated accordingly. 
6. Create lists from both the profile and lists page (with and without descriptions). Make sure that the modal disappears automatically and lists appear without refreshing.
7. Delete lists and lists page. Make sure that the modal disappears automatically and lists delete without refreshing.
8. Navigate to the contents of a single list. Edit List Details by changing the title and description.
9. Navigate to the contents of a single list. Delete songs and ensure that the list updates without refreshing.
10. Ensure that there is no access to edit, update, or delete the contents of another user's lists.

## ToDo - IMPORTANT FOR EVERYONE
- Add protection against editing lists to have no title.
- Button and overall design of components across the profile page require refactoring (Following vs Followers vs Reviews vs Lists) so they match
- No Albums on the Albums page need to show "No Albums" instead of "Loading"
- Modals for other components (following, albums, reviews) need to disappear automatically and the page must reflect the changes without refresh.
- Reviews should be able to be deleted (and possible editing if we have time)

blank
lists page
within a list

### CODE REVIEW

File: soundcrate/src/app/user/[username]/profile/page.js [hunk 1]
```diff
@@ -165,10 +178,17 @@ export default function UserProfilePage({ params }) {
       <section className="flex flex-col gap-2 grow shrink min-w-52 w-1/3">
```
Comment (annesatran) [inline soundcrate/src/app/user/[username]/profile/page.js:178]: Nit: Would be nice to have the message expand to full width like below:
image
This would require a bit of refactoring I think because the the change I made to get it to look like the picture (adding 'w-full' class and removing 'grid grid-col-4 gap-3' classes) messes up the format when there ARE albums to display

File: soundcrate/src/components/CreateListModal.js [hunk 2]
```diff
@@ -35,90 +34,73 @@ export default function CreateListModal({ username }) {
...
-        // get data from user
+
         const response_data = await response.json();
-        if (response_data?.status == 200) {
-          console.log(response_data.body)
-        } else if (response_data?.status == 401) {
-          // display unauthorized modal
-          HSOverlay?.close("#"+modal_id);
+        if (response.ok) {
+          onCreate(response_data.body);
```
Comment (annesatran) [inline soundcrate/src/components/CreateListModal.js]: The username is not getting passed into the newly created list, so the link doesn't work unless you refresh the page. It links to `/user/undefined/...` as the user:

image

The username could be passed into `onCreate` from here so that it works in any context `onCreate({username, ...response_data.body})`

File: soundcrate/src/app/user/[username]/profile/lists/page.js [hunk 3]
```diff
@@ -28,21 +30,61 @@ export default function ListsPage({ params }) {
     fetchSongLists(username);
   }, [username]);
 
+  // auto updates
+  const handleNewList = (newList) => {
+    setLists([newList, ...lists]);
```
Comment (annesatran) [inline soundcrate/src/app/user/[username]/profile/lists/page.js]: Can we update it so that the newList appears at the bottom, `[...lists, newList]`

File: soundcrate/src/app/api/lists/deleteSongList/route.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        }
+
+        await connectMongoDB();
+
+        const deletedList = await SongList.findByIdAndDelete(list_id);
+
+        if (!deletedList) {
+            return NextResponse.json({ message: "List not found" }, { status: 404 });
+        }
+
```
Comment (annesatran) [inline soundcrate/src/app/api/lists/deleteSongList/route.js:21]: Let's add an additional check here to make sure that the list actually belongs to the currently logged in user, like return a 401 Unauthorized error instead. This can also be for another PR to add middleware to protect the deleteSongList and editSongList routes.

File: soundcrate/src/app/api/lists/editSongList/route.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+                { status: 400 }
+            );
+        }
+
+        await connectMongoDB();
+
+        // Find the song list by ID and update it
+        const updatedList = await SongList.findByIdAndUpdate(
+            list_id,
+            {
```
Comment (annesatran) [inline soundcrate/src/app/api/lists/editSongList/route.js:22]: See message on deleteSongList route

File: soundcrate/src/app/user/[username]/profile/lists/page.js [hunk 6]
```diff
@@ -28,21 +30,61 @@ export default function ListsPage({ params }) {
...
+            song_count={list.songIds.length} 
+          />
+        </div>
+        {session?.status === 'authenticated' && session?.user?.username === username && (
+          <button
+            className="btn p-2 bg-dark-light text-white rounded-md hover:bg-red"
+            onClick={() => handleDeleteList(list._id)}
+          >
+            Delete
+          </button>
```
Comment (annesatran) [inline soundcrate/src/app/user/[username]/profile/lists/page.js:77]: Nit, flagging that a good future update would be to make this a component and have it on an individual song list page

**Review by annesatran — CHANGES_REQUESTED** [review]
Thank you for working on this!! Left mostly Nit comments for things we can change in a future update. Only real requested change is to fix the 'undefined' user link I described in my review, but otherwise looks good to merge!

File: soundcrate/src/app/user/[username]/profile/lists/page.js (hunk 6 above)
Comment (jemeduria) [reply soundcrate/src/app/user/[username]/profile/lists/page.js:77]: i'll add this idea to our slides instead of this merge for right now

File: soundcrate/src/app/api/lists/deleteSongList/route.js (hunk 4 above)
Comment (jemeduria) [reply soundcrate/src/app/api/lists/deleteSongList/route.js:21]: i'll add the need to more protections to our slides instead of this merge for right now, since the buttons only show up if its the user's account.

**Review by annesatran — COMMENTED** [review]
LGTM!
